File: sdk/android/src/main/java/com/tuya/smart/rnsdk/share/TuyaShareModule.java
package com.tuya.smart.rnsdk.share;

import com.alibaba.fastjson.JSON;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.tuya.smart.home.sdk.TuyaHomeSdk;
import com.tuya.smart.rnsdk.utils.JsonUtils;
import com.tuya.smart.rnsdk.utils.ReactParamsCheck;
import com.tuya.smart.rnsdk.utils.TuyaReactUtils;
import com.tuya.smart.sdk.bean.ShareIdBean;

import java.util.ArrayList;
import java.util.Arrays;

import javax.annotation.Nonnull;

import static com.tuya.smart.rnsdk.utils.Constant.AUTOSHARING;
import static com.tuya.smart.rnsdk.utils.Constant.COUNTRYCODE;
import static com.tuya.smart.rnsdk.utils.Constant.DEVID;
import static com.tuya.smart.rnsdk.utils.Constant.DEVIDS;
import static com.tuya.smart.rnsdk.utils.Constant.HOMEID;
import static com.tuya.smart.rnsdk.utils.Constant.MEMBERID;
import static com.tuya.smart.rnsdk.utils.Constant.NAME;
import static com.tuya.smart.rnsdk.utils.Constant.SHAREBEAN;
import static com.tuya.smart.rnsdk.utils.Constant.SHAREID;
import static com.tuya.smart.rnsdk.utils.Constant.USERACCOUNT;
import static com.tuya.smart.rnsdk.utils.Constant.coverDTL;
import static com.tuya.smart.rnsdk.utils.Constant.getIResultCallback;
import static com.tuya.smart.rnsdk.utils.Constant.getITuyaResultCallback;

public class TuyaShareModule extends ReactContextBaseJavaModule {
    public TuyaShareModule(@Nonnull ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Nonnull
    @Override
    public String getName() {
        return "TuyaShareModule";
    }

    @ReactMethod
    public void enableDevShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID, MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().enableDevShare(params.getString(DEVID),
                    coverDTL(params.getDouble(MEMBERID)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void disableDevShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID, MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().disableDevShare(params.getString(DEVID),
                    coverDTL(params.getDouble(MEMBERID)), getIResultCallback(promise));
        }
    }


    @ReactMethod
    public void removeReceivedDevShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeReceivedDevShare(params.getString(DEVID), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void addShareWithHomeId(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(HOMEID, COUNTRYCODE, USERACCOUNT, DEVIDS), params)) {
            ArrayList<String> list = new ArrayList<>();
            for (int i = 0; i < params.getArray(DEVIDS).size(); i++) {
                list.add(params.getArray(DEVIDS).getString(i));
            }
            TuyaHomeSdk.getDeviceShareInstance().addShareWithHomeId(
                    coverDTL(params.getDouble(HOMEID)),
                    params.getString(COUNTRYCODE),
                    params.getString(USERACCOUNT),
                    list,
                    getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void queryDevShareUserList(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryDevShareUserList(params.getString(DEVID),
                    getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void addShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(HOMEID, COUNTRYCODE, USERACCOUNT, SHAREBEAN, AUTOSHARING), params)) {
            TuyaHomeSdk.getDeviceShareInstance().addShare(
                    coverDTL(params.getDouble(HOMEID)),
                    params.getString(COUNTRYCODE),
                    params.getString(USERACCOUNT),
                    (ShareIdBean) JsonUtils.parse(JSON.toJSONString(TuyaReactUtils.parseToMap(params.getMap(SHAREBEAN))), ShareIdBean.class),
                    params.getBoolean(AUTOSHARING),
                    getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void addShareWithMemberId(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID, DEVIDS), params)) {
            ArrayList<String> list = new ArrayList<>();
            for (int i = 0; i < params.getArray(DEVIDS).size(); i++) {
                list.add(params.getArray(DEVIDS).getString(i));
            }
            TuyaHomeSdk.getDeviceShareInstance().addShareWithMemberId(coverDTL(params.getDouble(MEMBERID)),
                    list,
                    getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void getUserShareInfo(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().getUserShareInfo(
                    coverDTL(params.getDouble(MEMBERID)), getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void getReceivedShareInfo(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().getReceivedShareInfo(
                    coverDTL(params.getDouble(MEMBERID)), getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void queryUserShareList(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(HOMEID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryUserShareList(
                    coverDTL(params.getDouble(HOMEID)), getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void queryShareReceivedUserList(Promise promise) {
        TuyaHomeSdk.getDeviceShareInstance().queryShareReceivedUserList(getITuyaResultCallback(promise));
    }

    @ReactMethod
    public void queryShareDevFromInfo(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryShareDevFromInfo(params.getString(DEVID), getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void renameShareNickname(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID, NAME), params)) {
            TuyaHomeSdk.getDeviceShareInstance().renameShareNickname(
                    coverDTL(params.getDouble(MEMBERID)), params.getString(NAME), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void renameReceivedShareNickname(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID, NAME), params)) {
            TuyaHomeSdk.getDeviceShareInstance().renameReceivedShareNickname(
                    coverDTL(params.getDouble(MEMBERID)), params.getString(NAME), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void removeUserShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeUserShare(
                    coverDTL(params.getDouble(MEMBERID)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void removeReceivedUserShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeReceivedUserShare(
                    coverDTL(params.getDouble(MEMBERID)), getIResultCallback(promise));
        }
    }

    @ReactMethod
    public void onDestroy() {
        TuyaHomeSdk.getDeviceShareInstance().onDestroy();
    }


    @ReactMethod
    public void inviteShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(DEVID, USERACCOUNT, COUNTRYCODE), params)) {
            TuyaHomeSdk.getDeviceShareInstance().inviteShare(
                    params.getString(DEVID),
                    params.getString(USERACCOUNT),
                    params.getString(COUNTRYCODE), getITuyaResultCallback(promise));
        }
    }

    @ReactMethod
    public void confirmShareInviteShare(ReadableMap params, Promise promise) {
        if (ReactParamsCheck.checkParams(Arrays.asList(SHAREID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().confirmShareInviteShare(
                    params.getInt(SHAREID), getIResultCallback(promise));
        }
    }
}

File: android/src/main/java/com/tuya/smart/rnsdk/share/TuyaShareModule.kt
package com.tuya.smart.rnsdk.share

import com.facebook.react.bridge.*
import com.tuya.smart.home.sdk.TuyaHomeSdk
import com.tuya.smart.home.sdk.bean.ShareReceivedUserDetailBean
import com.tuya.smart.home.sdk.bean.ShareSentUserDetailBean
import com.tuya.smart.home.sdk.bean.SharedUserInfoBean
import com.tuya.smart.home.sdk.callback.ITuyaResultCallback
import com.tuya.smart.rnsdk.utils.Constant
import com.tuya.smart.rnsdk.utils.Constant.COUNTRYCODE
import com.tuya.smart.rnsdk.utils.Constant.DEVID
import com.tuya.smart.rnsdk.utils.Constant.DEVIDS
import com.tuya.smart.rnsdk.utils.Constant.HOMEID
import com.tuya.smart.rnsdk.utils.Constant.MEMBERID
import com.tuya.smart.rnsdk.utils.Constant.SHAREID
import com.tuya.smart.rnsdk.utils.Constant.NAME
import com.tuya.smart.rnsdk.utils.Constant.USERACCOUNT
import com.tuya.smart.rnsdk.utils.Constant.getIResultCallback
import com.tuya.smart.rnsdk.utils.JsonUtils
import com.tuya.smart.rnsdk.utils.ReactParamsCheck
import com.tuya.smart.rnsdk.utils.TuyaReactUtils


class TuyaShareModule(reactContext: ReactApplicationContext?) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "TuyaShareModule"
    }

    /*添加多个设备共享（追加）*/
    @ReactMethod
    fun addShareWithHomeId(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(HOMEID, COUNTRYCODE, USERACCOUNT, DEVIDS), params)) {
            var list = ArrayList<String>()
            var length = params.getArray(Constant.DEVIDS).size()
            for (index in 0..length) {
                list.add(params.getArray(Constant.DEVIDS).getString(index))
            }
            TuyaHomeSdk.getDeviceShareInstance().addShareWithHomeId(params.getDouble(HOMEID).toLong()
                    , params.getString(COUNTRYCODE),
                    params.getString(USERACCOUNT),
                    list,
                    getITuyaResultCallbackSingle(promise))
        }
    }

    /*批量添加设备共享*/
    @ReactMethod
    fun addShareWithMemberId(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID, DEVIDS), params)) {
            var list = ArrayList<String>()
            var length = params.getArray(Constant.DEVIDS).size()
            for (index in 0..length) {
                list.add(params.getArray(Constant.DEVIDS).getString(index))
            }
            TuyaHomeSdk.getDeviceShareInstance().addShareWithMemberId(params.getDouble(MEMBERID).toLong(),
                    list,
                    getIResultCallback(promise))
        }
    }
    /*单个设备取消共享*/
    @ReactMethod
    fun disableDevShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID, DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().disableDevShare(params.getString(DEVID), params.getDouble(MEMBERID).toLong(),
                    getIResultCallback(promise))
        }
    }
    /*查询主动分享的关系列表*/
    @ReactMethod
    fun queryUserShareList(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(HOMEID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryUserShareList(params.getDouble(HOMEID).toLong(),
                    getITuyaResultCallback(promise))
        }
    }
    /*查询收到分享关系列表*/
    @ReactMethod
    fun queryShareReceivedUserList(promise: Promise) {
        TuyaHomeSdk.getDeviceShareInstance().queryShareReceivedUserList(
                getITuyaResultCallback(promise))
    }
    /*查询指定设备的分享用户列表*/
    @ReactMethod
    fun queryDevShareUserList(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryDevShareUserList(params.getString(DEVID),
                    getITuyaResultCallback(promise))
        }
    }
    /* 查询指定设备是谁共享的 */
    @ReactMethod
    fun queryShareDevFromInfo(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().queryShareDevFromInfo(params.getString(DEVID),
                    getITuyaResultCallbackSingle(promise))
        }
    }
    /* 查询分享到指定用户的共享关系 */
    @ReactMethod
    fun getUserShareInfo(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().getUserShareInfo(
                    params.getDouble(MEMBERID).toLong(),
                    getITuyaResultShareSentUserDetailBeanCallback(promise))
        }
    }


    /*查询收到指定用户共享的信息*/
    @ReactMethod
    fun getReceivedShareInfo(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().getReceivedShareInfo(
                    params.getDouble(MEMBERID).toLong(),
                    object :  ITuyaResultCallback<ShareReceivedUserDetailBean> {
                        override fun onSuccess(bean: ShareReceivedUserDetailBean) {
                            promise.resolve(TuyaReactUtils.parseToWritableMap(bean))
                        }

                        override fun onError(errorCode: String, errorMessage: String) {
                            promise.reject(errorCode, errorMessage)
                        }
                    })
        }
    }


    /*可分享给未注册用户*/
    @ReactMethod
    fun inviteShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {

        }
    }

    /* 邀请分享确认*/
    @ReactMethod
    fun confirmShareInviteShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {

        }
    }


    /*删除共享关系*/
    @ReactMethod
    fun removeUserShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeUserShare(
                    params.getDouble(MEMBERID).toLong(),
                    getIResultCallback(promise))
        }
    }

    /*删除收到的共享关系*/
    @ReactMethod
    fun removeReceivedUserShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeReceivedUserShare(
                    params.getDouble(MEMBERID).toLong(),
                    getIResultCallback(promise))
        }
    }


    /*移除收到的分享设备*/
    @ReactMethod
    fun removeReceivedDevShare(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(DEVID), params)) {
            TuyaHomeSdk.getDeviceShareInstance().removeReceivedDevShare(
                    params.getString(DEVID),
                    getIResultCallback(promise))
        }
    }
    /*修改发出的分享人的备注名*/
    @ReactMethod
    fun renameShareNickname(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID, NAME), params)) {
            TuyaHomeSdk.getDeviceShareInstance().renameShareNickname(
                    params.getDouble(MEMBERID).toLong(),
                    params.getString(NAME),
                    getIResultCallback(promise))
        }
    }
    /*修改接收到的分享人的备注名*/
    @ReactMethod
    fun renameReceivedShareNickname(params: ReadableMap, promise: Promise) {
        if (ReactParamsCheck.checkParams(arrayOf(MEMBERID, NAME), params)) {
            TuyaHomeSdk.getDeviceShareInstance().renameReceivedShareNickname(
                    params.getDouble(MEMBERID).toLong(),
                    params.getString(NAME),
                    getIResultCallback(promise))
        }
    }

    fun getITuyaResultCallback(promise: Promise):  ITuyaResultCallback<List<SharedUserInfoBean>> {
        return object :  ITuyaResultCallback<List<SharedUserInfoBean>> {
            override fun onSuccess(p0: List<SharedUserInfoBean>) {
                promise.resolve(TuyaReactUtils.parseToWritableArray(
                        JsonUtils.toJsonArray(p0)))
            }

            override fun onError(p0: String?, p1: String?) {
                promise.reject(p0, p1)
            }
        }
    }

    fun getITuyaResultCallbackSingle(promise: Promise):  ITuyaResultCallback<SharedUserInfoBean> {
        return object :  ITuyaResultCallback<SharedUserInfoBean> {
            override fun onSuccess(bean: SharedUserInfoBean) {
                promise.resolve(TuyaReactUtils.parseToWritableMap(bean))
            }

            override fun onError(errorCode: String, errorMessage: String) {
                promise.reject(errorCode, errorMessage)
            }
        }
    }

    fun getITuyaResultShareSentUserDetailBeanCallback(promise: Promise):  ITuyaResultCallback<ShareSentUserDetailBean> {
        return object :  ITuyaResultCallback<ShareSentUserDetailBean> {
            override fun onSuccess(bean: ShareSentUserDetailBean) {
                promise.resolve(TuyaReactUtils.parseToWritableMap(bean))
            }

            override fun onError(errorCode: String, errorMessage: String) {
                promise.reject(errorCode, errorMessage)
            }
        }
    }
}